### PR TITLE
Add ability to override _reduce_ex_ function of DataPipe

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Required for `torchdata` graph functions

### CODE REVIEW

File: torch/utils/data/dataset.py [hunk 1]
```diff
@@ -185,6 +186,22 @@ def class_function(cls, source_dp, *args, **kwargs):
         function = functools.partial(class_function, cls_to_register)
         IterableDataset.functions[function_name] = function
 
+    def __reduce_ex__(self, *args, **kwargs):
+        print('called __reduce_ex', self)
+        if IterableDataset.reduce_ex_hook is not None:
+            print('called overridden __reduce_ex', self)
+            try:
+                return IterableDataset.reduce_ex_hook(self)
+            except NotImplementedError:
```
Comment (ejguan) [inline torch/utils/data/dataset.py:195]: Similar to the other PR, should it be a new Error?

File: torch/utils/data/dataset.py [hunk 2]
```diff
@@ -185,6 +186,22 @@ def class_function(cls, source_dp, *args, **kwargs):
...
+                pass
+        return super().__reduce_ex__(*args, **kwargs)
```
Comment (ejguan) [inline torch/utils/data/dataset.py:197]: TODO: we need to add doc for user to make sure all connected (source) `DataPipe`s are saved within `__dict__`.

File: torch/utils/data/dataset.py [hunk 3]
```diff
@@ -185,6 +186,22 @@ def class_function(cls, source_dp, *args, **kwargs):
         function = functools.partial(class_function, cls_to_register)
         IterableDataset.functions[function_name] = function
 
+    def __reduce_ex__(self, *args, **kwargs):
+        print('called __reduce_ex', self)
+        if IterableDataset.reduce_ex_hook is not None:
+            print('called overridden __reduce_ex', self)
```
Comment (ejguan) [inline torch/utils/data/dataset.py:192]: nit: remove `print`

**Review by ejguan — APPROVED** [review]
LGTM with some comments.

File: torch/utils/data/dataset.py (hunk 2 above)
Comment (VitalyFedyunin) [reply torch/utils/data/dataset.py:197]: I was thinking to have some sort of tests template, which will make sure that any pipe can be serialized, deserialized etc.

**Pull request closed without merging**